### PR TITLE
feat: support LOGIN auth mechanism

### LINKED ISSUE #14: Login Auth
https://github.com/umputun/remark42/issues/1472 

looks like Microsoft does not allow [Plain Auth anymore](https://support.microsoft.com/en-us/office/outlook-com-no-longer-supports-auth-plain-authentication-07f7d5e9-1697-465f-84d2-4513d4ff0145), which is used in [go-pkgz/email/email.go](https://github.com/go-pkgz/email/blob/master/email.go#L123).
The go `net/smtp` package does not support the login mechanism `LOGIN`.
Here is an example I found for login auth using `net/smtp` https://gist.github.com/homme/22b457eb054a07e7b2fb

### PR DESCRIPTION
**Pull request by gurza**
Closes #14

### CODE REVIEW

File: pkg/smtpauth/smtpauth.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// Implementation of additional SMTP authentication mechanisms.
```
Comment (umputun) [inline pkg/smtpauth/smtpauth.go]: I don't think this should nested package under pkg or even a separate package. I would probably won't even make a separate file for this thing, but rather would add to email.go

File: email.go [hunk 2]
```diff
@@ -50,6 +53,12 @@ type Params struct {
 	InlineImages    []string // InlineImages images path
 }
 
+// List of supported authentication methods
+const (
```
Comment (umputun) [inline email.go]: my suggestion is to use "fake enum", i.e custom type from string and a pair of values. This way we won't need to pass string as meth

File: email.go [hunk 3]
```diff
@@ -204,6 +213,24 @@ func (em *Sender) client() (c *smtp.Client, err error) {
...
 }
 
+// auth returns an smtp.Auth that implements SMTP authentication mechanism
+// depends on Sender settings.
+func (em *Sender) auth() smtp.Auth {
+	if em.smtpUserName == "" || em.smtpPassword == "" {
+		return nil // no auth
+	}
+
+	switch em.authMethod {
```
Comment (umputun) [inline email.go]: i don't think we need a switch here. A simple if authMethod == AuthMethodLogin will do just fine without a need in this switch and the dup code for the default

File: options.go [hunk 4]
```diff
@@ -62,6 +62,13 @@ func Auth(smtpUserName, smtpPasswd string) Option {
 	}
 }
 
+// AuthMethod sets auth method
+func AuthMethod(meth string) Option {
```
Comment (umputun) [inline options.go]: this can be simplified with LoginAuth options without any params. It will also make API easier to use

**Review by umputun — CHANGES_REQUESTED** [review]

**Comment by umputun** [comment]
I have added a few comments, mostly minor things. @paskal pls take a look too, thx

File: pkg/smtpauth/smtpauth.go (hunk 1 above)
Comment (gurza) [reply pkg/smtpauth/smtpauth.go]: I agree that using a nested package doesn't make sense for this small project. I'll move the code to a separate file. In the future, probably someone will want to add other authorization mechanisms, so I use a separate file.

File: email.go (hunk 2 above)
Comment (gurza) [reply email.go]: Agree

File: options.go (hunk 4 above)
Comment (gurza) [reply options.go]: LGTM, thanks for suggestion

File: email.go (hunk 3 above)
Comment (gurza) [reply email.go]: You are right. Currently, only two authentication methods are used. So for now, switch can be dropped.

**Comment by gurza** [comment]
@paskal hi! I have resolved all umputun comments. Please review it

File: options.go [hunk 5]
```diff
@@ -62,6 +62,13 @@ func Auth(smtpUserName, smtpPasswd string) Option {
 	}
 }
 
+// UseLoginAuth sets LOGIN auth method
+func UseLoginAuth() Option {
```
Comment (umputun) [inline options.go]: nit: it's better to drop Use prefix to be consistent with other options

Comment (gurza) [reply options.go]: @umputun 
It is not that simple.
Look. Standard `net/smtp` library contains `PlainAuth()` function for PLAIN auth. By analogy, I made `LoginAuth()` function for LOGIN auth. So I can't use `LoginAuth()` name in options.
Please recommend how it is better to name the option method.

**Comment by paskal** [comment]
I'll review it this evening. In the meantime, could you please run `golangci-lint run ./...` locally and fix the discovered errors?

File: options.go (hunk 5 above)
Comment (umputun) [reply options.go]: Unless I'm missing something, I don't see  how the LoginAuth in your code related to the isolated Option method's name

nm, i see what you mean

Comment (umputun) [reply options.go]: in this case, I'd rather call it `WithLoginAuth`

Comment (umputun) [reply options.go]: btw, any reason why your `LoginAuth` even needs to be public?

**Comment by gurza** [comment]
> I'll review it this evening. In the meantime, could you please run `golangci-lint run ./...` locally and fix the discovered errors?

I have resolved golangci-lint errors. thx

File: options.go (hunk 5 above)
Comment (gurza) [reply options.go]: There is no obvious need to export `LoginAuth()`. I propose to do the following.

1. Rename `LoginAuth()` (from `auth.go` file) to `newLoginAuth()` cuz this function is constructor for internal `loginAuth` struct that implements the smtp.Auth interface.

```go
func newLoginAuth(usr, pwd, host string) smtp.Auth {
	return &loginAuth{usr, pwd, host}
}

type loginAuth struct {
	user     string
	password string
	host     string
}
```

2. Use `LoginAuth()` name for option function.

@umputun  what do you think?

Comment (umputun) [reply options.go]: fine with me. Btw, not sure why this constructor is even needed, maybe just for aesthetic reasons/symmetry

Comment (gurza) [reply options.go]: It is needed more for aesthetic reasons.
It controls the set of required input parameters and cast the internal `loginAuth` type to `smtp.Auth` interface.

Comment (gurza) [reply options.go]: I have refactored `auth.go`.

1. Now all that is related to authentication is the internal types and functions.
2. The option that sets the LOGIN auth method is now called `LoginAuth` to be consistent with other options.

Comment (gurza) [reply options.go]: @umputun is it fine for you? May I resolve conversation?

**Comment by gurza** [comment]
@paskal hello!
Have you had a chance to review the code? Any additional comments?

**Review by paskal — APPROVED** [review]

**Comment by gurza** [comment]
@umputun let's merge? )

**Comment by umputun** [comment]
> @umputun let's merge? )

sure. Just two more things before we merge: 

1. Add info about new auth and a new option to the readme
2. rebase commits (squash)

**Comment by gurza** [comment]
@umputun now this feature branch is rebased and ready to be squashed and merged

**Comment by gurza** [comment]
@umputun I forgot to write right away, I added an info about new auth option to README.md. Looks like you can merge the PR.

**Pull request merged**